File: templates/python/code-carl/ch06_tree/t14_lc0513_find_bottom_left_value.py
"""LeetCode 513. Find Bottom Left Tree Value
https://leetcode.cn/problems/find-bottom-left-tree-value/
"""

from typing import Optional

from tree_node import TreeNode


class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        pass


if __name__ == "__main__":
    root1 = TreeNode(2, TreeNode(1), TreeNode(3))
    solution = Solution()
    assert solution.findBottomLeftValue(root1) == 1

    root2 = TreeNode(
        1, TreeNode(2, TreeNode(4)), TreeNode(3, TreeNode(5, TreeNode(7)), TreeNode(6))
    )
    solution = Solution()
    assert solution.findBottomLeftValue(root2) == 7

    print("All tests passed.")

File: solutions/java/code-carl/ch06_tree/T14_LC0513_FindBottomLeftValue.java
/**
 * Practice template for:
 * - LeetCode 513. Find Bottom Left Tree Value
 *   https://leetcode.cn/problems/find-bottom-left-tree-value/
 */
import java.util.ArrayDeque;
import java.util.Deque;

public class T14_LC0513_FindBottomLeftValue {

    public int findBottomLeftValue(TreeNode root) {
        if (root == null) {
            return -1;
        }

        int res = root.val;
        Deque<TreeNode> deque = new ArrayDeque<>();
        deque.offer(root);
        while (!deque.isEmpty()) {
            int size = deque.size();
            res = deque.peek().val;
            while (size -- > 0) {
                TreeNode curr = deque.poll();
                if (curr.left != null) {
                    deque.offer(curr.left);
                }
                if (curr.right != null) {
                    deque.offer(curr.right);
                }
            }
        }
        return res;
    }
}
